lib/iris/tests/unit/coords/test_AuxCoord.py

```diff
@@ -0,0 +1,764 @@
+        coord = AuxCoord(self.pts_real, bounds=self.bds_real)
+        result = coord.lazy_bounds()
+        self.assertIsInstance(result, da.Array)
+        self.assertArrayEqual(result.dtype, self.bds_real.dtype)
+        self.assertArrayEqual(result.compute(), self.bds_real)
+
+    def test_lazy_core(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_lazy)
+        result = coord.lazy_bounds()
+        self.assertIs(result, self.bds_lazy)
+
+
+@skip_lazy_auxcoords
+class Test_has_lazy_points(tests.IrisTest):
+    def setUp(self):
+        setup_test_arrays(self)
+
+    def test_real_core(self):
+        coord = AuxCoord(self.pts_real)
+        result = coord.has_lazy_points()
+        self.assertFalse(result)
+
+    def test_lazy_core(self):
+        coord = AuxCoord(self.pts_lazy)
+        result = coord.has_lazy_points()
+        self.assertTrue(result)
+
+    def test_lazy_core_realise(self):
+        coord = AuxCoord(self.pts_lazy)
+        coord.points
+        result = coord.has_lazy_points()
+        self.assertFalse(result)
+
+
+@skip_lazy_auxcoords
+class Test_has_lazy_bounds(tests.IrisTest):
+    def setUp(self):
+        setup_test_arrays(self)
+
+    def test_real_core(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_real)
+        result = coord.has_lazy_bounds()
+        self.assertFalse(result)
+
+    def test_lazy_core(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_lazy)
+        result = coord.has_lazy_bounds()
+        self.assertTrue(result)
+
+    def test_lazy_core_realise(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_lazy)
+        coord.bounds
+        result = coord.has_lazy_bounds()
+        self.assertFalse(result)
+
+
+@skip_lazy_auxcoords
+class Test_bounds_dtype(tests.IrisTest):
+    def test_i16(self):
+        test_dtype = np.int16
+        coord = AuxCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+    def test_u16(self):
+        test_dtype = np.uint16
+        coord = AuxCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+    def test_f16(self):
+        test_dtype = np.float16
+        coord = AuxCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+
+def lazyness_string(data):
+    return 'lazy' if isinstance(data, da.Array) else 'real'
+
+
+_DEBUG_ALL_TYPES = _DEBUG
+
+
+def coords_all_dtypes_and_lazynesses(self, dtypes=(np.float64, np.int16)):
+    # Generate coords with all possible types of points and bounds, and all
+    # of the given dtypes.
+    points_types = ['real']
+    bounds_types = ['no', 'real']
+    if _AUXCOORD_LAZY_API:
+        points_types.append('lazy')
+        bounds_types.append('lazy')
+    for dtype in (np.float64, np.int16):
```

Comment:
It feels like you could compress all these `for` loops neatly using functionality provided by `itertools`. I'm sure this has been done in tests recently, but I couldn't find an example quickly...